updater/osv/osv.go

```diff
@@ -0,0 +1,801 @@
+				if err != nil {
+					zlog.Warn(ctx).Err(err).Msg("event version error")
+				}
+			}
+			var vs string
+			switch r.Type {
+			case `ECOSYSTEM`:
+				vs = b.String()
+			}
+			pkg, novel := e.LookupPackage(af.Package.PURL, vs)
+			v.Package = pkg
+			if novel {
+				pkg.RepositoryHint = af.Package.Ecosystem
+			}
+			if repo := e.LookupRepository(name); repo != nil {
+				v.Repo = repo
+			}
+
+		}
+	}
+	return nil
+}
+
+// All the methods follow the same pattern: just reslice the slice if
+// there's space, or use append to do an alloc+copy.
+
+func (e *ecs) NewVulnerability() *claircore.Vulnerability {
+	i := len(e.Vulnerability)
+	if cap(e.Vulnerability) > i {
+		e.Vulnerability = e.Vulnerability[:i+1]
+	} else {
+		e.Vulnerability = append(e.Vulnerability, claircore.Vulnerability{})
+	}
+	return &e.Vulnerability[i]
+}
+
+func (e *ecs) LookupPackage(name string, ver string) (*claircore.Package, bool) {
+	e.hash.Reset()
+	fmt.Fprintf(e.hash, "%s\x00%s", name, ver) // documented to never error
+	key := e.hash.Sum64()
+	i, ok := e.pkgindex[key]
+	if !ok {
+		i = len(e.Package)
+		if cap(e.Package) > i {
+			e.Package = e.Package[:i+1]
+		} else {
+			e.Package = append(e.Package, claircore.Package{})
+		}
+		e.Package[i].Name = name
+		e.Package[i].Version = ver
+		e.pkgindex[key] = i
+	}
+	return &e.Package[i], ok
+}
+
+func (e *ecs) LookupRepository(name string) (r *claircore.Repository) {
+	e.hash.Reset()
+	fmt.Fprint(e.hash, name) // documented to never error
+	key := e.hash.Sum64()
+	i, ok := e.repoindex[key]
+	if !ok {
+		i = len(e.Repository)
+		if cap(e.Repository) > i {
+			e.Repository = e.Repository[:i+1]
+		} else {
+			e.Repository = append(e.Repository, claircore.Repository{})
+		}
+		e.Repository[i].Name = name
+		switch name {
+		case "crates.io":
+			e.Repository[i].URI = `https://crates.io/`
+		case "go":
+			e.Repository[i].URI = `https://pkg.go.dev/`
+		case "npm":
+			e.Repository[i].URI = `https://www.npmjs.com/`
+		case "nuget":
+			e.Repository[i].URI = `https://www.nuget.org/packages/`
+		case "oss-fuzz":
+			e.Repository[i].URI = `https://google.github.io/oss-fuzz/`
+		case "packagist":
+			e.Repository[i].URI = `https://packagist.org/`
+		case "pypi":
+			e.Repository[i].URI = `https://pypi.org/`
+		case "rubygems":
+			e.Repository[i].URI = `https://rubygems.org/gems/`
+		case "maven":
+			e.Repository[i].URI = `https://maven.apache.org/repository/`
+		}
+		e.repoindex[key] = i
+	}
+	return &e.Repository[i]
+}
+
+func (e *ecs) Len() int {
+	return len(e.Vulnerability)
+}
+
+func (e *ecs) Finalize() []*claircore.Vulnerability {
+	r := make([]*claircore.Vulnerability, len(e.Vulnerability))
+	for i := range e.Vulnerability {
+		r[i] = &e.Vulnerability[i]
+	}
+	return r
+}
+
+// FromCVSS is an attempt at an implementation of the scale and formulas
+// described here: https://www.first.org/cvss/v3.1/specification-document#Qualitative-Severity-Rating-Scale
+func fromCVSS(s string) (sev claircore.Severity, err error) {
```

Comment:
out of interest, did you see this: https://github.com/goark/go-cvss? Might be interesting to run [this test suite](https://github.com/goark/go-cvss/blob/master/v3/base/base_test.go) through this function, specifically:

```		{vector: "CVSS:3.1/AV:P/AC:H/PR:H/UI:R/S:U/C:N/I:N/A:N", score: 0.0, severity: SeverityNone},     //Zero metrics
		{vector: "CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:U/C:H/I:N/A:N", score: 7.5, severity: SeverityHigh},     //CVE-2015-8252
		{vector: "CVSS:3.1/AV:N/AC:L/PR:N/UI:R/S:C/C:L/I:L/A:N", score: 6.1, severity: SeverityMedium},   //CVE-2013-1937
		{vector: "CVSS:3.1/AV:N/AC:L/PR:L/UI:N/S:C/C:L/I:L/A:N", score: 6.4, severity: SeverityMedium},   //CVE-2013-0375
		{vector: "CVSS:3.1/AV:N/AC:H/PR:N/UI:R/S:U/C:L/I:N/A:N", score: 3.1, severity: SeverityLow},      //CVE-2014-3566
		{vector: "CVSS:3.1/AV:N/AC:L/PR:L/UI:N/S:C/C:H/I:H/A:H", score: 9.9, severity: SeverityCritical}, //CVE-2012-1516
		{vector: "CVSS:3.1/AV:N/AC:L/PR:L/UI:N/S:U/C:H/I:H/A:H", score: 8.8, severity: SeverityHigh},     //CVE-2012-0384
		{vector: "CVSS:3.1/AV:L/AC:L/PR:N/UI:R/S:U/C:H/I:H/A:H", score: 7.8, severity: SeverityHigh},     //CVE-2015-1098
		{vector: "CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:U/C:H/I:N/A:N", score: 7.5, severity: SeverityHigh},     //CVE-2014-0160
		{vector: "CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:U/C:H/I:H/A:H", score: 9.8, severity: SeverityCritical}, //CVE-2014-6271
		{vector: "CVSS:3.1/AV:N/AC:H/PR:N/UI:N/S:C/C:N/I:H/A:N", score: 6.8, severity: SeverityMedium},   //CVE-2008-1447
		{vector: "CVSS:3.1/AV:P/AC:L/PR:N/UI:N/S:U/C:H/I:H/A:H", score: 6.8, severity: SeverityMedium},   //CVE-2014-2005
		{vector: "CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:C/C:L/I:N/A:N", score: 5.8, severity: SeverityMedium},   //CVE-2010-0467
		{vector: "CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:C/C:N/I:L/A:N", score: 5.8, severity: SeverityMedium},   //CVE-2012-1342
		{vector: "CVSS:3.1/AV:N/AC:L/PR:L/UI:R/S:C/C:L/I:L/A:N", score: 5.4, severity: SeverityMedium},   //CVE-2014-9253
		{vector: "CVSS:3.1/AV:L/AC:L/PR:N/UI:R/S:U/C:H/I:H/A:H", score: 7.8, severity: SeverityHigh},     //CVE-2009-0658
		{vector: "CVSS:3.1/AV:A/AC:L/PR:N/UI:N/S:U/C:H/I:H/A:H", score: 8.8, severity: SeverityHigh},     //CVE-2011-1265
		{vector: "CVSS:3.1/AV:P/AC:L/PR:N/UI:N/S:U/C:N/I:H/A:N", score: 4.6, severity: SeverityMedium},   //CVE-2014-2019
		{vector: "CVSS:3.1/AV:N/AC:L/PR:N/UI:R/S:U/C:H/I:H/A:H", score: 8.8, severity: SeverityHigh},     //CVE-2015-0970
		{vector: "CVSS:3.1/AV:N/AC:H/PR:N/UI:N/S:U/C:H/I:H/A:N", score: 7.4, severity: SeverityHigh},     //CVE-2014-0224
		{vector: "CVSS:3.1/AV:N/AC:L/PR:N/UI:R/S:C/C:H/I:H/A:H", score: 9.6, severity: SeverityCritical}, //CVE-2012-5376
```
Just a thought